pkg/controllers/nodeclaim/disruption/underutilization.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disruption
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/controllers/state"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+	nodeclaimutil "sigs.k8s.io/karpenter/pkg/utils/nodeclaim"
+)
+
+// Underutilization is a nodeclaim sub-controller that adds or removes status conditions on empty nodeclaims based on TTLSecondsAfterUnderutilized
+type Underutilization struct {
+	kubeClient client.Client
+	cluster    *state.Cluster
+}
+
+//nolint:gocyclo
+func (e *Underutilization) Reconcile(ctx context.Context, nodePool *v1.NodePool, nodeClaim *v1.NodeClaim) (reconcile.Result, error) {
+	hasUnderutilizedCondition := nodeClaim.StatusConditions().Get(v1.ConditionTypeUnderutilized) != nil
+
+	// 1. If ConsolidationPolicyWhenUnderutilized is not configured or ConsolidateAfter isn't configured, remove the underutilization status condition
+	if nodePool.Spec.Disruption.ConsolidationPolicy != v1.ConsolidationPolicyWhenUnderutilized ||
+		nodePool.Spec.Disruption.ConsolidateAfter == nil ||
```

Comment:
I think we should make it so that this struct can't be nil and only the duration can set to be nil (set to `Never`)